### PR TITLE
chore: deprecate juju.jasyncio

### PR DESCRIPTION
**Pull request by EdmilsonRodrigues**
#### Description

*\<I created a new file called _jasyncio, passed all functions from jasyncio to there, but maintained the jasyncio file with a deprecation warning. Then, I changed jasyncio to _jasyncio or asyncio in all files that it appeared.\>*

*\<Fixes: Stopped using jasyncio and imported asynchronous functions directly from asyncio\>*

#### QA Steps

*\<Commands / tests / steps to run to verify that the change works: I do not know yet, this is my first contribution to an open source project, and I could use some guidance.\>*

```
tox -e py3 -- tests/unit/...
```

```
tox -e integration -- tests/integration/...
```

All CI tests need to pass.

*\<Please note that most likely an additional test will be required by the reviewers for any change that's not a one liner to land.\>*

#### Notes & Discussion

*\<This is my first time contributing and I am still learning how. Can you help me telling how can I test if the changes work, if I should do more smaller commits or only big commits, or any other guidance of how to work on this project?\>*

### CODE REVIEW

**Comment by dimaqq** [comment]
Thank you for the pull request!

There's a tonne of changes stemming from `jasyncio.run()` which seems similar to `asyncio.run()` in spirit with a couple of caveats:
- custom signal handler
- the `steps` concept

Would you consider taking some time to analyse if these two caveats are actually important?

My gut tells me that it should be safe to drop this custom crud and use `asyncio.run()` instead.

File: tests/unit/test_proxy.py [hunk 1]
```diff
@@ -37,16 +39,18 @@ def test_proxy_from_config_non_arg(self):
...
-            "config": {
-                "api-host": "https://localhost:8456",
-                "namespace": "controller-python-test",
-                "remote-port": "1234",
-                "service": "controller",
-                "service-account-token": "==AA",
-                "ca-cert": "==AA",
-            },
-        })
+        proxy = proxy_from_config(
```
Comment (dimaqq) [inline tests/unit/test_proxy.py]: Please prefer pre-commit over other tools.

This change seems to be the result of `ruff format`, while the original was the result of `ruff format --preview`.

File: juju/loop.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 import warnings
 
 warnings.warn(
-    "juju.loop module is being deprecated by 3.0, use juju.jasyncio instead",
+    "juju.loop module is being deprecated by 3.0, use juju._jasyncio instead",
```
Comment (dimaqq) [inline juju/loop.py:9]: this file can be removed instead

File: juju/application.py [hunk 3]
```diff
@@ -207,10 +208,12 @@ async def add_unit(self, count=1, to=None, attach_storage=[]):
...
-        return await jasyncio.gather(*[
-            jasyncio.ensure_future(self.model._wait_for_new("unit", unit_id))
-            for unit_id in result.units
-        ])
+        return await asyncio.gather(
+            *[
+                asyncio.ensure_future(self.model._wait_for_new("unit", unit_id))
+                for unit_id in result.units
+            ]
+        )
```
Comment (dimaqq) [inline juju/application.py]: asyncio.gather already converts coroutines to tasks, there's no need to call .ensure_future.

let's simplify this.

File: juju/model.py [hunk 4]
```diff
@@ -1900,10 +1901,14 @@ async def deploy(
...
-                    for app_name in pending_apps
-                ])
+                await asyncio.gather(
+                    *[
+                        asyncio.ensure_future(
+                            self._wait_for_new("application", app_name)
+                        )
+                        for app_name in pending_apps
+                    ]
+                )
```
Comment (dimaqq) [inline juju/model.py]: same here, ensure_future is not needed.

**Comment by dimaqq** [comment]
Re CLA: head to launchpad.net to sign the contributor license agreement. it's the same as for Ubuntu.

**Comment by dimaqq** [comment]
Fixes https://github.com/juju/python-libjuju/issues/1178

Also requires rebase.

**Comment by dimaqq** [comment]
The RTD fails because the docs still reference `juju.loop`.
Let's remove juju.loop entirely.
I don't think jasyncio is even documented, so that's nice :)

File: tests/unit/test_loop.py [hunk 5]
```diff
@@ -1,37 +1,30 @@
...
 import unittest
 
-from juju import jasyncio
-
 
 class TestLoop(unittest.TestCase):
     def setUp(self):
         # new event loop for each test
-        policy = jasyncio.get_event_loop_policy()
+        policy = asyncio.get_event_loop_policy()
```
Comment (dimaqq) [inline tests/unit/test_loop.py:11]: This file can now be removed, as there's no point in testing Python stdlib.

File: tests/integration/test_unit.py [hunk 6]
```diff
@@ -164,7 +164,7 @@ def check_results(results, out):
 async def test_scp():
     # ensure that asyncio.subprocess will work;
     try:
-        asyncio.get_child_watcher().attach_loop(jasyncio.get_running_loop())
+        asyncio.get_child_watcher().attach_loop(asyncio.get_running_loop())
```
Comment (dimaqq) [inline tests/integration/test_unit.py]: get_child_watcher is deprecated in Python 3.12.
Let's refactor this... somehow, I think it's safe to remove these 4 lines.
We're in control of how the integration test are run, I would imagine that pytest-asyncio uses the main thread by default, so the check if pointless.

File: juju/jasyncio.py [hunk 7]
```diff
@@ -82,6 +83,13 @@
 
 import websockets
 
+warnings.warn(
+    "juju.jasyncio module is being deprecated by 3.0, use asyncio or juju._jasyncio instead",
+    DeprecationWarning,
+    stacklevel=2,
+)
+
+
```
Comment (dimaqq) [inline juju/jasyncio.py:89]: I think this file should re-export the useful _jasyncio functions instead of having own copies.

Only the `run` function should remain defined in this file.

(asyncio re-exports are fine).

**Comment by dimaqq** [comment]
The CLA check still fails...
https://ubuntu.com/legal/contributors is where the CLA can be signed.

Just in case the CLA checker is confused about the commit authorship, here's the log from the failed check:

```
Run canonical/has-signed-canonical-cla@1.2.3
Installing python3-launchpadlib

Checking the following users on GitHub:
Check in the signed list service
- EdmilsonRodrigues ✕ (has not signed the CLA)

Checking the following user on Launchpad:
- e.rodrigues@roborian.com ✕ (has no Launchpad account)
```

**Comment by dimaqq** [comment]
Maybe there's a mismatch between email registered on launchpad and the email in the commits' author field?

**Comment by dimaqq** [comment]
There are still some pre-commit issues... please run `pre-commit -a` in your local checkout, it should show all the errors and/or file modifications.

**Comment by dimaqq** [comment]
Unit tests fail on Python 3.8 due to type hint `Task[Any]`.

If you use `pyright`, I think it should warn you correctly about python level.
Usually what we do is add `from __future__ import annotations` at the top of each file.

**Comment by EdmilsonRodrigues** [comment]
> The CLA check still fails... https://ubuntu.com/legal/contributors is where the CLA can be signed.
> 
> Just in case the CLA checker is confused about the commit authorship, here's the log from the failed check:
> 
> ```
> Run canonical/has-signed-canonical-cla@1.2.3
> Installing python3-launchpadlib
> 
> Checking the following users on GitHub:
> Check in the signed list service
> - EdmilsonRodrigues ✕ (has not signed the CLA)
> 
> Checking the following user on Launchpad:
> - e.rodrigues@roborian.com ✕ (has no Launchpad account)
> ```

Sorry about this. Just trying to sign the CLA again. Also fixed the other issues and am committing now, thank you for the pacience.

File: juju/model.py [hunk 8]
```diff
@@ -1398,7 +1398,8 @@ async def _notify_observers(self, delta, old_obj, new_obj):
 
         for o in self._observers:
             if o.cares_about(delta):
-                asyncio.ensure_future(o(delta, old_obj, new_obj, self))
+                f = asyncio.ensure_future(o(delta, old_obj, new_obj, self))
+                f.add_done_callback(lambda f: f.result())
```
Comment (dimaqq) [inline juju/model.py]: I'd rather this hack was not removed.
It's most likely there to prevent "Task was never awaited for" warnings.
Not a great practice, but out of scope of this PR.

**Comment by dimaqq** [comment]
One of the commits was not signed correctly, please rebase and re-sign that commit.
P.S. the issue is probably wrong author email (?)

Screenshot 2024-12-05 at 17 57 32
Screenshot 2024-12-05 at 17 57 25

**Comment by dimaqq** [comment]
Hi Edmilson, there's something wrong with your branch history: `+18,814 −692`

It seems that your branch includes the upstream changes, like "add 3.5.5 schema" and "release 3.6.0.0" without a merge commit. Thus, GitHub considers a boatload of changes as unique to this branch, for example the type hints on "live_map".

I'm not sure what's the best approach, generally there are two:
- a. either rebase your branch on top of latest main (and fix conflicts correctly)
- b. or merge latest main into your branch (and fix conflicts correctly)

I prefer rebase over merge, but sometimes it's really just picking the lesser evil.

Finally if that proves to be too hard, there's always this approach:
- make a diff between your current branch and main
- create a new branch off main, apply that diff
- the result should contain all of your changes and only your changes.

**Comment by EdmilsonRodrigues** [comment]
Sorry about this. As I said, this is my first time contributing for an open source project.

I think I fixed it this time.

File: tests/unit/test_loop.py (hunk 5 above)
Comment (dimaqq) [reply tests/unit/test_loop.py:11]: I suppose I can always do that in a separate PR too.

File: tests/charm-secret/pyproject.toml [hunk 9]
```diff
@@ -18,7 +18,7 @@ target-version = ["py38"]
 [tool.ruff]
 line-length = 99
 select = ["E", "W", "F", "C", "N", "D", "I001"]
-extend-ignore = [
+lint.extend-ignore = [
```
Comment (dimaqq) [inline tests/charm-secret/pyproject.toml:21]: Apparently that too is deprecated, and it's just "ignore" now,,,
Totally OK to leave to a separate PR.
https://docs.astral.sh/ruff/settings/#lint_extend-ignore

File: uv.lock [hunk 10]
```diff
@@ -0,0 +1,779 @@
+version = 1
+requires-python = ">=3.8"
+
+[[package]]
+name = "bcrypt"
+version = "4.2.1"
```
Comment (dimaqq) [inline uv.lock]: We should not check in the lock file.
There are pros and cons to lock files for libraries, though at any rate, it should not be part of this PR.

File: juju/utils.py [hunk 11]
```diff
@@ -60,7 +62,7 @@ def juju_config_dir():
     return str(config_dir.expanduser().resolve())
 
 
-def juju_ssh_key_paths():
+def juju_ssh_key_paths() -> tuple[str, str]:
```
Comment (dimaqq) [inline juju/utils.py:65]: Unrelated changes, but welcome anyway.

File: juju/utils.py [hunk 12]
```diff
@@ -139,7 +145,9 @@ async def _block():
 
 
 async def block_until_with_coroutine(
-    condition_coroutine, timeout=None, wait_period=0.5
+    condition_coroutine: Callable[[], Coroutine[Any, Any, bool]],
```
Comment (dimaqq) [inline juju/utils.py]: What do you think about a type hint in this style instead?

```suggestion
    condition_coroutine: Callable[[], Awaitable[bool]],
```

**Comment by dimaqq** [comment]
Looking good, I've kicked off the tests!

**Review by dimaqq — APPROVED** [review]
I think it looks good!

I'll ask for a 2nd pair of eyes.

File: juju/loop.py (hunk 2 above)
Comment (benhoyt) [reply juju/loop.py:9]: Should we just say "use asyncio instead"?

**Review by benhoyt — APPROVED** [review]
This looks like good work to me, thanks @EdmilsonRodrigues! One nit question about a warning message.

File: juju/loop.py (hunk 2 above)
Comment (dimaqq) [reply juju/loop.py:9]: True, but I'm going to remove `juju/loop.py` entirely in a separate PR.
